core-primitives/stf-state-handler/src/state_handler.rs

```diff
@@ -25,144 +25,261 @@ use crate::{
 	error::{Error, Result},
 	handle_state::HandleState,
 	query_shard_state::QueryShardState,
+	state_initializer::InitializeState,
 	state_snapshot_repository::VersionedStateAccess,
 };
+use itp_hashing::Hash;
+use itp_sgx_externalities::SgxExternalitiesTrait;
 use itp_stf_state_observer::traits::UpdateState;
 use itp_types::ShardIdentifier;
-use std::{sync::Arc, vec::Vec};
+use std::{collections::HashMap, sync::Arc, vec::Vec};
+
+type StatesMap<State, Hash> = HashMap<ShardIdentifier, (State, Hash)>;
 
 /// Implementation of the `HandleState` trait.
 ///
-/// It's a concurrency wrapper around a state snapshot repository, which handles
-/// access to any shards and state files. The state handler ensures we have thread-safe
-/// concurrent access to that repository.
-pub struct StateHandler<Repository, StateObserver> {
+/// Responsible for handling any state instances. Holds a map with all the latest states for each shard.
+/// In addition, uses the snapshot repository to save file snapshots of a state.
+pub struct StateHandler<Repository, StateObserver, StateInitializer>
+where
+	Repository: VersionedStateAccess,
+{
 	state_snapshot_repository: RwLock<Repository>,
+	states_map_lock: RwLock<StatesMap<Repository::StateType, Repository::HashType>>,
 	state_observer: Arc<StateObserver>,
+	state_initializer: Arc<StateInitializer>,
 }
 
-impl<Repository, StateObserver> StateHandler<Repository, StateObserver> {
-	pub fn new(state_snapshot_repository: Repository, state_observer: Arc<StateObserver>) -> Self {
+impl<Repository, StateObserver, StateInitializer>
+	StateHandler<Repository, StateObserver, StateInitializer>
+where
+	Repository: VersionedStateAccess,
+	Repository::StateType: Hash<Repository::HashType>,
+	StateObserver: UpdateState<Repository::StateType>,
+	StateInitializer: InitializeState<StateType = Repository::StateType>,
+{
+	/// Creates a new instance WITHOUT loading any state from the repository.
+	/// Results in an empty states map.
+	pub fn new(
+		state_snapshot_repository: Repository,
+		state_observer: Arc<StateObserver>,
+		state_initializer: Arc<StateInitializer>,
+	) -> Self {
+		Self::new_with_states_map(
+			state_snapshot_repository,
+			state_observer,
+			state_initializer,
+			Default::default(),
+		)
+	}
+
+	/// Create a new state handler and initialize its state map with the
+	/// states that are available in the snapshot repository.
+	pub fn load_from_repository(
+		state_snapshot_repository: Repository,
+		state_observer: Arc<StateObserver>,
+		state_initializer: Arc<StateInitializer>,
+	) -> Result<Self> {
+		let states_map = Self::load_all_latest_snapshots(&state_snapshot_repository)?;
+		Ok(Self::new_with_states_map(
+			state_snapshot_repository,
+			state_observer,
+			state_initializer,
+			states_map,
+		))
+	}
```

Comment:
Creating a new state handler is the only case where we read from our state file snapshot repository. We read all the latest states for each shard and populate our internal state map with it.